File: lulu/binary_tree_preorder_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Preorder in ArrayList which contains node values.
    """
    def preorderTraversal(self, root):
        # write your code here
        result = []
        self.pre_order_traversal(root, result)
        return result

    def pre_order_traversal(self, root, result):
    	if root is None:
    		return
    	result.append(root.val)
    	self.pre_order_traversal(root.left, result)
    	self.pre_order_traversal(root.right, result)

File: bsdelf/binary-tree-preorder-traversal.cc
/**
 * @author      Yanhui Shen
 * @version     1.0
 * @since       2016-02-11 
 */

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in vector which contains node values.
     */
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> vals;
        deque<TreeNode*> nodes = { root };
        while (!nodes.empty()) {
            const auto node = nodes.front();
            nodes.pop_front();
            if (not node) continue;
            vals.push_back(node->val);
            nodes.push_front(node->right);
            nodes.push_front(node->left);
        }
        return vals;
    }
};
